### PR TITLE
Implement generic bitstring for primitive types

### PR DESCRIPTION
**Pull request by quinnj**
I've recently found myself defining a few primitive types, and it's
handy to have `bitstring` work generically instead of having to cast to
`BitInteger`s. This definition is also nice since it works on primitive
types larger than `UInt128`. The inner loop vectorized body is copied from the
`bin` function.

cc: @nickrobinson251, since we chatted briefly about this

### CODE REVIEW

File: base/intfuncs.jl [hunk 1]
```diff
@@ -805,13 +805,23 @@ julia> bitstring(2.2)
...
-bitstring(x::Union{Char,Int32,UInt32,Float32}) = string(reinterpret(UInt32,x), pad = 32, base = 2)
-bitstring(x::Union{Int64,UInt64,Float64})      = string(reinterpret(UInt64,x), pad = 64, base = 2)
-bitstring(x::Union{Int128,UInt128})            = string(reinterpret(UInt128,x), pad = 128, base = 2)
+function bitstring(x::T) where {T}
+    isprimitivetype(T) || throw(ArgumentError("$T not a primitive type"))
+    sz = sizeof(T) * 8
+    str = Base.StringVector(sz)
+    i = sz
+    @inbounds while i >= 4
+        b = UInt32(Base.trunc_int(UInt8, x))
```
Comment (vtjnash) [inline base/intfuncs.jl:814]: this is illegal if `x` is already 8 bits or smaller
```
julia> Base.trunc_int(UInt8, 0x10)
ERROR: Trunc: output bitsize must be < input bitsize
```

Comment (quinnj) [reply base/intfuncs.jl:814]: Ah good catch; we don't allow primitive types < 8 bits, right?

File: base/intfuncs.jl [hunk 2]
```diff
@@ -805,13 +805,23 @@ julia> bitstring(2.2)
...
-bitstring(x::Union{Char,Int32,UInt32,Float32}) = string(reinterpret(UInt32,x), pad = 32, base = 2)
-bitstring(x::Union{Int64,UInt64,Float64})      = string(reinterpret(UInt64,x), pad = 64, base = 2)
-bitstring(x::Union{Int128,UInt128})            = string(reinterpret(UInt128,x), pad = 128, base = 2)
+function bitstring(x::T) where {T}
+    isprimitivetype(T) || throw(ArgumentError("$T not a primitive type"))
+    sz = sizeof(T) * 8
+    str = Base.StringVector(sz)
+    i = sz
+    @inbounds while i >= 4
+        b = UInt32(x isa UInt8 ? x : Base.trunc_int(UInt8, x))
```
Comment (KristofferC) [inline base/intfuncs.jl:814]: There is still an error here if `x` is of size 8, no?

```jl
julia> Base.trunc_int(UInt8, Int8(3))
ERROR: Trunc: output bitsize must be < input bitsize
   ```

Comment (quinnj) [reply base/intfuncs.jl]: Yep, good catch. Just pushed a fix.

**Comment by quinnj** [comment]
linuxppc64le failure seems unrelated (seems like some build-related failure)

**Pull request merged**